ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import static java.nio.file.Files.deleteIfExists;
+
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.TransientQueryMetadata;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.StreamsConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TransientQueryCleanupService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(TransientQueryCleanupService.class);
+  private static final Pattern TRANSIENT_PATTERN =
+          Pattern.compile("(?i).*transient_.*_[0-9]\\d*_[0-9]\\d*");
+  private static final int NUM_OF_RETRIES_TO_DELETE_RESOURCE = 5;
+
+  private final BlockingQueue<Callable<Boolean>> cleanupTasks;
+  private final Retryer<Boolean> retryer;
+  private final String stateDir;
+  private final ServiceContext serviceContext;
+  private final Set<TransientQueryMetadata> allTransientQueriesEverCreated;
```

Comment:
We probably shouldn't keep a reference to the metadata directly since we'll never GC any of these objects or the streams topologies they point to, so it's essentially a memory leak.  Doing that for a set of Strings is maybe tolerable in practice, but we probably should't do this.

Reply:
good idea. Changed to query the QueryRegistry to check the status of the query (running or not) instead of maintaining a set of all the queries ever created. This will avoid the memory leak